File: jqurantree/encoding.py
"""
Encoding — Unicode, Buckwalter, Simple codecs with mapping tables.
"""

from __future__ import annotations

import unicodedata

from .types import CharacterType, DiacriticTypes, ArabicCharacter

# ── Unicode mapping tables ──

CP_TO_CHAR: dict[int, CharacterType] = {
    0x0627: CharacterType.ALIF, 0x0628: CharacterType.BA, 0x062A: CharacterType.TA,
    0x062B: CharacterType.THA, 0x062C: CharacterType.JEEM, 0x062D: CharacterType.HHA,
    0x062E: CharacterType.KHA, 0x062F: CharacterType.DAL, 0x0630: CharacterType.THAL,
    0x0631: CharacterType.RA, 0x0632: CharacterType.ZAIN, 0x0633: CharacterType.SEEN,
    0x0634: CharacterType.SHEEN, 0x0635: CharacterType.SAD, 0x0636: CharacterType.DAD,
    0x0637: CharacterType.TTA, 0x0638: CharacterType.ZZA, 0x0639: CharacterType.AIN,
    0x063A: CharacterType.GHAIN, 0x0641: CharacterType.FA, 0x0642: CharacterType.QAF,
    0x0643: CharacterType.KAF, 0x0644: CharacterType.LAM, 0x0645: CharacterType.MEEM,
    0x0646: CharacterType.NOON, 0x0647: CharacterType.HA, 0x0648: CharacterType.WAW,
    0x0649: CharacterType.ALIF_MAKSURA, 0x064A: CharacterType.YA,
    0x0621: CharacterType.HAMZA, 0x0622: CharacterType.ALIF_MADDA,
    0x0623: CharacterType.ALIF_HAMZA_ABOVE, 0x0624: CharacterType.WAW_HAMZA,
    0x0625: CharacterType.ALIF_HAMZA_BELOW, 0x0626: CharacterType.YA_HAMZA,
    0x0629: CharacterType.TA_MARBUTA, 0x0640: CharacterType.TATWEEL,
    0x0671: CharacterType.ALIF_WASLA,
    0x06E5: CharacterType.SMALL_WAW, 0x06E6: CharacterType.SMALL_YA,
    0x06E8: CharacterType.SMALL_HIGH_NOON, 0x06ED: CharacterType.SMALL_LOW_MEEM,
    0x06DC: CharacterType.HIGH_SEEN, 0x06DF: CharacterType.HIGH_ROUNDED_ZERO,
    0x06E0: CharacterType.SMALL_HIGH_UPRIGHT_RECT_ZERO,
    0x06EA: CharacterType.EMPTY_CENTRE_LOW_STOP,
    0x06EB: CharacterType.EMPTY_CENTRE_HIGH_STOP,
    0x06EC: CharacterType.ROUNDED_HIGH_STOP_FILLED,
}

CP_TO_DIACRITIC: dict[int, DiacriticTypes] = {
    0x064E: DiacriticTypes.FATHA, 0x064F: DiacriticTypes.DAMMA,
    0x0650: DiacriticTypes.KASRA, 0x0651: DiacriticTypes.SHADDA,
    0x0652: DiacriticTypes.SUKUN,
    0x064B: DiacriticTypes.TANWEEN_FATHA, 0x064C: DiacriticTypes.TANWEEN_DAMMA,
    0x064D: DiacriticTypes.TANWEEN_KASRA,
}

CHAR_TO_CP: dict[CharacterType, int] = {v: k for k, v in CP_TO_CHAR.items()}

DIACRITIC_TO_CP: dict[DiacriticTypes, int] = {
    DiacriticTypes.FATHA: 0x064E, DiacriticTypes.DAMMA: 0x064F,
    DiacriticTypes.KASRA: 0x0650, DiacriticTypes.SHADDA: 0x0651,
    DiacriticTypes.SUKUN: 0x0652,
    DiacriticTypes.TANWEEN_FATHA: 0x064B, DiacriticTypes.TANWEEN_DAMMA: 0x064C,
    DiacriticTypes.TANWEEN_KASRA: 0x064D,
}

DIACRITIC_ORDER = [DiacriticTypes.SHADDA, DiacriticTypes.FATHA, DiacriticTypes.DAMMA,
                    DiacriticTypes.KASRA, DiacriticTypes.TANWEEN_FATHA,
                    DiacriticTypes.TANWEEN_DAMMA, DiacriticTypes.TANWEEN_KASRA,
                    DiacriticTypes.SUKUN]

DIACRITIC_CPS = frozenset(CP_TO_DIACRITIC.keys()) | {0x0670}
ALIF_KHANJAREEYA_CP = 0x0670


# ── Decode / Encode ──

def decode(text: str, normalize: bool = True) -> list[ArabicCharacter]:
    if not text: return []
    if normalize: text = unicodedata.normalize("NFC", text)
    result: list[ArabicCharacter] = []
    i, n = 0, len(text)
    while i < n:
        cp = ord(text[i])
        if cp in DIACRITIC_CPS and result:
            if cp == ALIF_KHANJAREEYA_CP:
                result.append(ArabicCharacter(CharacterType.ALIF_KHANJAREEYA))
            elif cp in CP_TO_DIACRITIC:
                prev = result[-1]
                result[-1] = ArabicCharacter(prev.char_type, prev.diacritics | CP_TO_DIACRITIC[cp])
            i += 1
        elif cp in CP_TO_CHAR:
            result.append(ArabicCharacter(CP_TO_CHAR[cp]))
            i += 1
        else:
            i += 1
    return result


def encode(characters: list[ArabicCharacter]) -> str:
    parts: list[str] = []
    for ac in characters:
        ct = ac.char_type
        if ct == CharacterType.ALIF_KHANJAREEYA:
            parts.append(chr(ALIF_KHANJAREEYA_CP))
            continue
        base = CHAR_TO_CP.get(ct)
        if base is not None: parts.append(chr(base))
        for dt in DIACRITIC_ORDER:
            if ac.diacritics & dt:
                dc = DIACRITIC_TO_CP.get(dt)
                if dc is not None: parts.append(chr(dc))
    return "".join(parts)


# ── Buckwalter ──

_CHAR_TO_BW: dict[CharacterType, str] = {
    CharacterType.HAMZA: "'", CharacterType.ALIF_MADDA: "|",
    CharacterType.ALIF_HAMZA_ABOVE: ">", CharacterType.WAW_HAMZA: "&",
    CharacterType.ALIF_HAMZA_BELOW: "<", CharacterType.YA_HAMZA: "}",
    CharacterType.ALIF: "A", CharacterType.BA: "b", CharacterType.TA_MARBUTA: "p",
    CharacterType.TA: "t", CharacterType.THA: "v", CharacterType.JEEM: "j",
    CharacterType.HHA: "H", CharacterType.KHA: "x", CharacterType.DAL: "d",
    CharacterType.THAL: "*", CharacterType.RA: "r", CharacterType.ZAIN: "z",
    CharacterType.SEEN: "s", CharacterType.SHEEN: "$", CharacterType.SAD: "S",
    CharacterType.DAD: "D", CharacterType.TTA: "T", CharacterType.ZZA: "Z",
    CharacterType.AIN: "E", CharacterType.GHAIN: "g", CharacterType.FA: "f",
    CharacterType.QAF: "q", CharacterType.KAF: "k", CharacterType.LAM: "l",
    CharacterType.MEEM: "m", CharacterType.NOON: "n", CharacterType.HA: "h",
    CharacterType.WAW: "w", CharacterType.ALIF_MAKSURA: "Y", CharacterType.YA: "y",
    CharacterType.ALIF_WASLA: "{", CharacterType.TATWEEL: "_",
    CharacterType.ALIF_KHANJAREEYA: "`",
}

_DIACRITIC_TO_BW: dict[DiacriticTypes, str] = {
    DiacriticTypes.FATHA: "a", DiacriticTypes.DAMMA: "u",
    DiacriticTypes.KASRA: "i", DiacriticTypes.SUKUN: "o",
    DiacriticTypes.SHADDA: "~", DiacriticTypes.TANWEEN_FATHA: "F",
    DiacriticTypes.TANWEEN_DAMMA: "N", DiacriticTypes.TANWEEN_KASRA: "K",
}

_BW_REVERSE: dict[str, CharacterType] = {v: k for k, v in _CHAR_TO_BW.items()}
_BW_DIAC_REVERSE: dict[str, DiacriticTypes] = {v: k for k, v in _DIACRITIC_TO_BW.items()}


def encode_buckwalter(characters: list[ArabicCharacter]) -> str:
    parts: list[str] = []
    for ac in characters:
        bw = _CHAR_TO_BW.get(ac.char_type, "")
        parts.append(bw)
        for dt in DIACRITIC_ORDER:
            if ac.diacritics & dt:
                bd = _DIACRITIC_TO_BW.get(dt, "")
                if bd: parts.append(bd)
    return "".join(parts)


def decode_buckwalter(text: str) -> list[ArabicCharacter]:
    result: list[ArabicCharacter] = []
    i = 0
    while i < len(text):
        ch = text[i]
        if ch in _BW_REVERSE:
            result.append(ArabicCharacter(_BW_REVERSE[ch]))
        elif result and ch in _BW_DIAC_REVERSE:
            prev = result[-1]
            result[-1] = ArabicCharacter(prev.char_type, prev.diacritics | _BW_DIAC_REVERSE[ch])
        i += 1
    return result


# ── Simple encoder ──

_STRIP_MAP: dict[CharacterType, CharacterType] = {
    CharacterType.ALIF_HAMZA_ABOVE: CharacterType.ALIF,
    CharacterType.ALIF_HAMZA_BELOW: CharacterType.ALIF,
    CharacterType.ALIF_WASLA: CharacterType.ALIF,
    CharacterType.ALIF_MADDA: CharacterType.ALIF,
    CharacterType.ALIF_MAKSURA: CharacterType.YA,
    CharacterType.WAW_HAMZA: CharacterType.WAW,
    CharacterType.YA_HAMZA: CharacterType.YA,
    CharacterType.TA_MARBUTA: CharacterType.HA,
}

_STRIP_SKIP = frozenset({
    CharacterType.ALIF_KHANJAREEYA, CharacterType.TATWEEL,
    CharacterType.SMALL_WAW, CharacterType.SMALL_YA,
    CharacterType.SMALL_HIGH_NOON, CharacterType.SMALL_LOW_MEEM,
    CharacterType.HIGH_SEEN, CharacterType.HIGH_ROUNDED_ZERO,
    CharacterType.SMALL_HIGH_UPRIGHT_RECT_ZERO,
    CharacterType.EMPTY_CENTRE_LOW_STOP, CharacterType.EMPTY_CENTRE_HIGH_STOP,
    CharacterType.ROUNDED_HIGH_STOP_FILLED,
})


def encode_simple(characters: list[ArabicCharacter]) -> list[ArabicCharacter]:
    result: list[ArabicCharacter] = []
    for ac in characters:
        ct = ac.char_type
        if ct in _STRIP_SKIP: continue
        stripped = _STRIP_MAP.get(ct, ct)
        result.append(ArabicCharacter(stripped, DiacriticTypes.NONE))
    return result

File: QuranNative/Sources/QuranKit/Encoding.swift
import Foundation

public let CP_TO_CHAR_TYPE: [UInt32: CharacterType] = [
    0x0621: .hamza,
    0x0622: .alifWithMadda,
    0x0623: .alifWithHamzaAbove,
    0x0624: .wawWithHamza,
    0x0625: .alifWithHamzaBelow,
    0x0626: .yaWithHamza,
    0x0627: .alif,
    0x0628: .ba,
    0x0629: .taMarbuta,
    0x062A: .ta,
    0x062B: .tha,
    0x062C: .jeem,
    0x062D: .hha,
    0x062E: .kha,
    0x062F: .dal,
    0x0630: .thal,
    0x0631: .ra,
    0x0632: .zai,
    0x0633: .seen,
    0x0634: .sheen,
    0x0635: .sad,
    0x0636: .dad,
    0x0637: .tta,
    0x0638: .zza,
    0x0639: .ain,
    0x063A: .ghain,
    0x0641: .fa,
    0x0642: .qaf,
    0x0643: .kaf,
    0x0644: .lam,
    0x0645: .meem,
    0x0646: .noon,
    0x0647: .ha,
    0x0648: .waw,
    0x0649: .alifMaksura,
    0x064A: .ya,
    0x0671: .alifWasla,
    0x0640: .tatweel,
    0x0670: .daggerAlif,
    0x06D6: .smallHighSadLamAlef,
    0x06D7: .smallHighQafLamAlef,
    0x06D8: .smallHighMeemInit,
    0x06D9: .smallHighLamAlef,
    0x06DA: .smallHighJeem,
    0x06DB: .smallHighThreeDots,
    0x06DC: .smallHighSeen,
    0x06DF: .smallHighRoundedZero,
    0x06E0: .smallHighRectZero,
    0x06E2: .smallHighMeem,
    0x06E3: .smallLowSeen,
    0x06E5: .smallWaw,
    0x06E6: .smallYeh,
    0x06E7: .smallHighYeh,
    0x06E8: .smallHighNoon,
    0x06EA: .emptyCentreLowStop,
    0x06EB: .emptyCentreHighStop,
    0x06EC: .roundedHighStop,
    0x06ED: .smallLowMeem,
]

public let CP_TO_DIACRITIC: [UInt32: DiacriticTypes] = [
    0x064E: .fatha,
    0x064F: .damma,
    0x0650: .kasra,
    0x0651: .shadda,
    0x0652: .sukun,
    0x064B: .tanweenFatha,
    0x064C: .tanweenDamma,
    0x064D: .tanweenKasra,
    0x0653: .maddahAbove,
    0x0654: .hamzaAbove,
]

public let CHAR_TYPE_TO_BUCKWALTER: [CharacterType: Character] = [
    .hamza: "'", .alifWithMadda: "|", .alifWithHamzaAbove: ">",
    .wawWithHamza: "&", .alifWithHamzaBelow: "<", .yaWithHamza: "}",
    .alif: "A", .ba: "b", .taMarbuta: "p", .ta: "t",
    .tha: "v", .jeem: "j", .hha: "H", .kha: "x",
    .dal: "d", .thal: "*", .ra: "r", .zai: "z",
    .seen: "s", .sheen: "$", .sad: "S", .dad: "D",
    .tta: "T", .zza: "Z", .ain: "E", .ghain: "g",
    .fa: "f", .qaf: "q", .kaf: "k", .lam: "l",
    .meem: "m", .noon: "n", .ha: "h", .waw: "w",
    .alifMaksura: "Y", .ya: "y", .alifWasla: "{",
    .tatweel: "_", .daggerAlif: "`",
]

public let CHAR_TYPE_TO_CP: [CharacterType: UnicodeScalar] = [
    .hamza: "\u{0621}", .alifWithMadda: "\u{0622}", .alifWithHamzaAbove: "\u{0623}",
    .wawWithHamza: "\u{0624}", .alifWithHamzaBelow: "\u{0625}", .yaWithHamza: "\u{0626}",
    .alif: "\u{0627}", .ba: "\u{0628}", .taMarbuta: "\u{0629}", .ta: "\u{062A}",
    .tha: "\u{062B}", .jeem: "\u{062C}", .hha: "\u{062D}", .kha: "\u{062E}",
    .dal: "\u{062F}", .thal: "\u{0630}", .ra: "\u{0631}", .zai: "\u{0632}",
    .seen: "\u{0633}", .sheen: "\u{0634}", .sad: "\u{0635}", .dad: "\u{0636}",
    .tta: "\u{0637}", .zza: "\u{0638}", .ain: "\u{0639}", .ghain: "\u{063A}",
    .fa: "\u{0641}", .qaf: "\u{0642}", .kaf: "\u{0643}", .lam: "\u{0644}",
    .meem: "\u{0645}", .noon: "\u{0646}", .ha: "\u{0647}", .waw: "\u{0648}",
    .alifMaksura: "\u{0649}", .ya: "\u{064A}", .alifWasla: "\u{0671}",
    .tatweel: "\u{0640}", .daggerAlif: "\u{0670}",
    .smallHighSadLamAlef: "\u{06D6}", .smallHighQafLamAlef: "\u{06D7}",
    .smallHighMeemInit: "\u{06D8}", .smallHighLamAlef: "\u{06D9}",
    .smallHighJeem: "\u{06DA}", .smallHighThreeDots: "\u{06DB}",
    .smallHighSeen: "\u{06DC}", .smallHighRoundedZero: "\u{06DF}",
    .smallHighRectZero: "\u{06E0}", .smallHighMeem: "\u{06E2}",
    .smallLowSeen: "\u{06E3}", .smallWaw: "\u{06E5}",
    .smallYeh: "\u{06E6}", .smallHighYeh: "\u{06E7}",
    .smallHighNoon: "\u{06E8}", .emptyCentreLowStop: "\u{06EA}",
    .emptyCentreHighStop: "\u{06EB}", .roundedHighStop: "\u{06EC}",
    .smallLowMeem: "\u{06ED}",
]

public let DIACRITIC_TO_BUCKWALTER: [DiacriticTypes: Character] = [
    .fatha: "a", .damma: "u", .kasra: "i",
    .shadda: "~", .sukun: "o",
    .tanweenFatha: "F", .tanweenDamma: "N", .tanweenKasra: "K",
]

public let BUCKWALTER_TO_CP: [Character: UnicodeScalar] = {
    var result: [Character: UnicodeScalar] = [:]
    for (ct, bw) in CHAR_TYPE_TO_BUCKWALTER {
        if let cp = CHAR_TYPE_TO_CP[ct] {
            result[bw] = cp
        }
    }
    result["a"] = "\u{064E}"
    result["u"] = "\u{064F}"
    result["i"] = "\u{0650}"
    result["o"] = "\u{0652}"
    result["~"] = "\u{0651}"
    result["F"] = "\u{064B}"
    result["N"] = "\u{064C}"
    result["K"] = "\u{064D}"
    return result
}()

private let DIACRITIC_CP_ORDER: [UInt32] = [
    0x0651, 0x064E, 0x064F, 0x0650,
    0x064B, 0x064C, 0x064D, 0x0652,
    0x0653, 0x0654,
]

public func decode(_ text: String) -> [ArabicCharacter] {
    var result: [ArabicCharacter] = []
    for scalar in text.precomposedStringWithCanonicalMapping.unicodeScalars {
        let cp = scalar.value
        if let charType = CP_TO_CHAR_TYPE[cp] {
            result.append(ArabicCharacter(charType: charType))
        } else if let diacritic = CP_TO_DIACRITIC[cp] {
            if var last = result.last {
                last.diacritics.insert(diacritic)
                result[result.count - 1] = last
            }
        }
    }
    return result
}

public func encode(_ characters: [ArabicCharacter]) -> String {
    var scalars = String.UnicodeScalarView()
    for ch in characters {
        if let cp = CHAR_TYPE_TO_CP[ch.charType] {
            scalars.append(cp)
        }
        for cpOrder in DIACRITIC_CP_ORDER {
            switch cpOrder {
            case 0x0651: if ch.diacritics.contains(.shadda) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x064E: if ch.diacritics.contains(.fatha) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x064F: if ch.diacritics.contains(.damma) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x0650: if ch.diacritics.contains(.kasra) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x064B: if ch.diacritics.contains(.tanweenFatha) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x064C: if ch.diacritics.contains(.tanweenDamma) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x064D: if ch.diacritics.contains(.tanweenKasra) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x0652: if ch.diacritics.contains(.sukun) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x0653: if ch.diacritics.contains(.maddahAbove) { scalars.append(UnicodeScalar(cpOrder)!) }
            case 0x0654: if ch.diacritics.contains(.hamzaAbove) { scalars.append(UnicodeScalar(cpOrder)!) }
            default: break
            }
        }
    }
    return String(scalars)
}

public func decodeBuckwalter(_ text: String) -> [ArabicCharacter] {
    var result: [ArabicCharacter] = []
    var i = text.startIndex
    while i < text.endIndex {
        let ch = text[i]
        if let cp = BUCKWALTER_TO_CP[ch] {
            if let charType = CP_TO_CHAR_TYPE[cp.value] {
                result.append(ArabicCharacter(charType: charType))
            } else if let diacritic = CP_TO_DIACRITIC[cp.value] {
                if var last = result.last {
                    last.diacritics.insert(diacritic)
                    result[result.count - 1] = last
                }
            }
        }
        i = text.index(after: i)
    }
    return result
}
